oscore-cf/src/main/java/org/eclipse/californium/oscore/ObjectSecurityLayer.java

```diff
@@ -0,0 +1,309 @@
+	 * 
+	 * @param message  the message
+	 * @param ctx  the OSCore context
+	 * 
+	 * @return  the encrypted message
+	 * 
+	 * @throws OSException error while encrypting request
+	 */
+	public static Request prepareSend(Request message, OSCoreCtx ctx, int realCode) 
+			throws OSException {
+		return RequestEncryptor.encrypt(message, ctx, realCode);
+	}
+
+	/**
+	 * Encrypt an outgoing response using the OSCore context.
+	 * 
+	 * @param message  the message
+	 * @param ctx  the OSCore context
+	 * 
+	 * @return  the encrypted message
+	 * 
+	 * @throws OSException error while encrypting response
+	 */
+	public static Response prepareSend(Response message, OSCoreCtx ctx, int realCode, final boolean newPartialIV) 
+			throws OSException {
+		return ResponseEncryptor.encrypt(message, ctx, realCode, newPartialIV);
+	}
+
+	/**
+	 * Decrypt an incoming request using the right OSCore context
+	 * 
+	 * @param request  the incoming request
+	 * @param db  the OSCore context database
+	 * 
+	 * @return  the cid of the OSCore context used for this request
+	 * 
+	 * @throws OSException  error while decrypting request
+	 */
+	public static byte[] prepareReceive(Request request) 
+			throws OSException {
+		return RequestDecryptor.decrypt(request);
+	}
+
+	/**
+	 * Decrypt an incoming response using the right OSCore context
+	 * 
+	 * @param response  the incoming request
+	 * @param db  the OSCore context database
+	 * @return  the decrypted and verified response
+	 * 
+	 * @throws OSException error while decrypting response
+	 */
+	public static Response prepareReceive(Response response) 
+			throws OSException {
+		return ResponseDecryptor.decrypt(response);
+	}
+
+	@Override
+	public void sendRequest(Exchange exchange, Request request){
+		Request req = request;
+		if(shouldProtectRequest(request)){
+			try {
+				String uri = request.getURI();             
+				OSCoreCtxDB db = HashMapCtxDB.getInstance();
+				
+				if(uri == null){
+					LOGGER.severe(OSExceptionHandler.URI_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.URI_NULL);
+				}
+				if(db == null){
+					LOGGER.severe(OSExceptionHandler.DB_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.DB_NULL);
+				}
+				
+				OSCoreCtx ctx = db.getContext(uri);
+				if( ctx == null){
+					LOGGER.severe(OSExceptionHandler.CTX_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.CTX_NULL);
+				}
+
+				exchange.setCryptographicContextID(ctx.getRecipientId());
+				int seqByToken = ctx.getSenderSeq();
+				int realCode = request.getCode().value;				
+				request = OptionJuggle.setFakeCodeRequest(request);
+				
+				req = prepareSend(request, ctx, realCode);
+				if (req.getToken() != null) {//Save the context by token to speed up lookup
+					db.addContext(request.getToken(), ctx);
+					db.addSeqByToken(request.getToken(), seqByToken);
+
+				} else {
+
+					Token token = db.generateToken();
```

Comment:
At that point, the payload encryption is done?
The token is therefore only created to be used as key?
If so, try to add a `MessageObserver` and implement the `db.add???` in the `onReadyToSend` callback using the then create token from the request and remove your own token generator.

Reply:
Yes and yes. Sounds like a great idea.

Reply:
Suggestion: if the request arrives at this point without a token been set by the CoapStack I will throw an exception and discontinue processing of request. Then I get rid of token generation.

Reply:
This suggestion will not work. I will look into your suggestion instead.